### PR TITLE
Allow always block peer with unexpected fingerprint.

### PR DESCRIPTION
**Pull request by StageGuard**

### CODE REVIEW

File: app/shared/app-data/src/commonMain/kotlin/data/models/preference/TorrentPeerConfig.kt [hunk 1]
```diff
@@ -28,6 +28,11 @@ data class TorrentPeerConfig(
     val idRegexFilters: List<String> = emptyList(),
     val enableIdFilter: Boolean = false,
 
+    /**
+     * [PeerInfo.id] 一般是以 ASCII `-xxxxxx-` 开头的 ByteArray. 例如 Ani 的是 `-aniLT3000-`.
+     * 如果 Peer 连接的客户端不是以此开头, 可以将其视为异常客户端.
+     */
+    val blockUnexpectedId: Boolean = false,
```
Comment (Him188) [inline app/shared/app-data/src/commonMain/kotlin/data/models/preference/TorrentPeerConfig.kt:35]: 默认 true?

File: app/shared/app-data/src/commonMain/kotlin/tools/torrent/PeerFilterType.kt [hunk 2]
```diff
@@ -18,10 +18,23 @@ class PeerClientFilter(patternRegex: String) : PeerFilter {
     }
 }
 
+object PeerUnexpectedIdFilter : PeerFilter {
+    private const val FINGERPRINT_TRIM = '-'
+            
+    override fun onFilter(info: PeerInfo): Boolean {
```
Comment (Him188) [inline app/shared/app-data/src/commonMain/kotlin/tools/torrent/PeerFilterType.kt:24]: 这 filter 怎么是返回 `true` 剔除, 跟 `List.filter` 行为不一致, 反习惯

Comment (Him188) [reply app/shared/app-data/src/commonMain/kotlin/tools/torrent/PeerFilterType.kt:24]: 但不要在这个 PR 里解决这个问题

File: torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt [hunk 3]
```diff
@@ -10,9 +10,7 @@ class SwigPeerInfo(
     native: peer_info_t,
 ) : PeerInfo {
     override val handle: HandleId = native.torrent_handle_id
-    override val id: String = buildString {
-        native.peer_id.forEach { append(it) }
-    }
+    override val id: CharArray = native.peer_id.toCharArray()
```
Comment (Him188) [inline torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt:13]: C++ char 实际上对应 Kotlin byte

File: app/shared/ui-settings/src/commonMain/kotlin/ui/settings/tabs/media/torrent/peer/PeerFilterSettingsState.kt [hunk 4]
```diff
@@ -58,12 +46,18 @@ class PeerFilterSettingsState(
         default = ""
     )
     
-    var idFilterEnabled by storage.prop({ it.enableIdFilter }, { copy(enableIdFilter = it) }, false)
+    var idFilterEnabled by storage.prop(
+        get = { it.enableIdFilter }, 
+        // 启用的时候 blockUnexpectedId 一定默认不启用, 禁用的时候 blockUnexpectedId 一定被禁用.
+        copy = { copy(enableIdFilter = it, blockUnexpectedId = false) },
```
Comment (Him188) [inline app/shared/ui-settings/src/commonMain/kotlin/ui/settings/tabs/media/torrent/peer/PeerFilterSettingsState.kt:52]: 改一个设置的时候不应该副作用修改别的设置

应该在你引擎逻辑里总的判断一下 idFilterEnabled

File: app/shared/ui-settings/src/commonMain/kotlin/ui/settings/tabs/media/torrent/peer/PeerFilterEditPane.kt [hunk 5]
```diff
@@ -31,56 +31,77 @@ import me.him188.ani.app.ui.richtext.RichText
...
+            { BBCodeSupportingText(supportingTextBBCode) }
+        },
+        trailingContent = { Switch(checked = enabled, onCheckedChange = onSwitchChange) },
+        colors = listItemColors,
+        modifier = Modifier.clickable { onSwitchChange(!enabled) },
+    )
+}
 
+@Composable
+private fun ExpandableSwitchItem(
```
Comment (Him188) [inline app/shared/ui-settings/src/commonMain/kotlin/ui/settings/tabs/media/torrent/peer/PeerFilterEditPane.kt:59]: 以后有这种修改先直接 commit 到 main, 不要放到 PR 里产生 (许多) 无关变更

File: app/shared/app-data/src/commonMain/kotlin/data/models/preference/TorrentPeerConfig.kt (hunk 1 above)
Comment (Him188) [inline app/shared/app-data/src/commonMain/kotlin/data/models/preference/TorrentPeerConfig.kt:35]: "expect" 的语义不合适吧

我们并没有 expect 什么. 用 invalid 比较合适?

File: app/shared/app-data/src/commonMain/kotlin/tools/torrent/PeerFilterType.kt [hunk 6]
```diff
@@ -18,10 +18,23 @@ class PeerClientFilter(patternRegex: String) : PeerFilter {
...
+        val idArr = info.id
+        // id 开头不是 fingerprint 就拒绝连接 
+        if (idArr[0] != FINGERPRINT_TRIM) return true
+        if (!idArr.drop(1).contains(FINGERPRINT_TRIM)) return true
```
Comment (Him188) [inline app/shared/app-data/src/commonMain/kotlin/tools/torrent/PeerFilterType.kt:28]: drop 会 copy 一次, 造成内存浪费

用 lastIndexOf, 判断 != 0

**Review by Him188 — CHANGES_REQUESTED** [review]

File: torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt (hunk 3 above)
Comment (Him188) [reply torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt:13]: Key Differences
Encoding and Size:

C++ char is 1 byte (ASCII), while Kotlin Char is 2 bytes (UTF-16).
Type Representation:

C++ char is more closely related to an integer, while Kotlin Char is a distinct type with additional properties and methods.
Character Handling:

Kotlin Char has built-in methods for common character checks (e.g., isLetter(), isWhitespace()), whereas in C++ you typically rely on functions like isalpha() from <cctype>.
String Handling:

C++ uses null-terminated character arrays for strings, whereas Kotlin uses immutable String objects.

Comment (StageGuard) [reply torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt:13]: 那我觉得应该也没问题，c++ char to kotlin char 也不会产生错误，也方便做判断

Comment (Him188) [reply torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt:13]: 它是byte，用char表示要涉及编码，会显著慢而且占位置

Comment (StageGuard) [reply torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt:13]: 那就没办法通过 regex 匹配过滤

Comment (Him188) [reply torrent/anitorrent/src/jvmMain/kotlin/session/SwigPeerInfo.kt:13]: ok fine

**Review by Him188 — APPROVED** [review]

**Pull request merged**